test/e2e/observer/e2e_test.go

```diff
@@ -9,26 +9,33 @@ import (
 	"path"
 	"path/filepath"
 	"regexp"
+	"strings"
 	"testing"
 
+	"github.com/openshift/ci-tools/pkg/steps/multi_stage"
 	"github.com/openshift/ci-tools/test/e2e/framework"
 )
 
 const (
 	defaultJobSpec = `JOB_SPEC={"type":"presubmit","job":"pull-ci-test-test-master-success","buildid":"0","prowjobid":"uuid","refs":{"org":"test","repo":"test","base_ref":"master","base_sha":"6d231cc37652e85e0f0e25c21088b73d644d89ad","pulls":[{"number":1234,"author":"droslean","sha":"538680dfd2f6cff3b3506c80ca182dcb0dd22a58"}]},"decoration_config":{"timeout":"4h0m0s","grace_period":"30m0s","utility_images":{"clonerefs":"registry.ci.openshift.org/ci/clonerefs:latest","initupload":"registry.ci.openshift.org/ci/initupload:latest","entrypoint":"registry.ci.openshift.org/ci/entrypoint:latest","sidecar":"registry.ci.openshift.org/ci/sidecar:latest"},"resources":{"clonerefs":{"limits":{"memory":"3Gi"},"requests":{"cpu":"100m","memory":"500Mi"}},"initupload":{"limits":{"memory":"200Mi"},"requests":{"cpu":"100m","memory":"50Mi"}},"place_entrypoint":{"limits":{"memory":"100Mi"},"requests":{"cpu":"100m","memory":"25Mi"}},"sidecar":{"limits":{"memory":"2Gi"},"requests":{"cpu":"100m","memory":"250Mi"}}},"gcs_configuration":{"bucket":"origin-ci-test","path_strategy":"single","default_org":"openshift","default_repo":"origin","mediaTypes":{"log":"text/plain"}},"gcs_credentials_secret":"gce-sa-credentials-gcs-publisher"}}`
 )
 
+var (
+	observerHasFailedRegexp = regexp.MustCompile("Step (.+-observer) failed after")
```

Comment:
This is smelly. Can you elaborate more on why we need to parse the error log to understand which step failed?

Reply:
I have noticed (and you can find some logs backing up this assumption [here](https://issues.redhat.com/browse/DPTP-3140)) that sometime the test gets flaky because the observer didn't run at at all. This is more or less what could happen:
- the [test step](https://github.com/openshift/ci-tools/blob/28deee345611ee06adf123ed9736263005add134/test/e2e/observer/failing-observer.yaml#L15) starts
- the [observer](https://github.com/openshift/ci-tools/blob/28deee345611ee06adf123ed9736263005add134/test/e2e/observer/failing-observer.yaml#L25) is supposed to run concurrently but it never gets the chance to. That is a mere timing issue not related with the code
-  the `e2e` test expects the observer to have logged its failure but it finds none
- the whole test fails but it's just a flake

What I'm trying to do here (without doing any invasive change to the `ci-operator` code) is to check for the line the observer is supposed to have logged; if there is none then it means the previous scenario happened and we can therefore ignore it

Reply:
I don't have a clear picture of how the observers work, but if we run the observer in parallel, we should have a watcher that monitors the observer's behavior. The fact that the test awaits for error and finds none, is a bug and not a flake. We should fix that instead.

Reply:
If an observer fails, the `ci-operator` should not consider the whole test to be failed, this is by design and this `e2e` test try to ensure this behavior holds by looking at the logs there were produced. The observer is ephemeral by nature, the `ci-operator` run it at best effort and, from what I have understood, in some cases the whole tests chain executes but the observer(s) does not.